---
title: "3. Exploratory) variables & PCA & UMAP"
author: "Yujin L"
date: "`r Sys.Date()`"
output: html_document
---

## 1 Load packages

Besides Seurat, we need to import some useful packages.

dplyr and ggplot2 are part of tidyverse, a readable and flexible R language for solving data science challenges. I personally prefer the coding style with tidyverse, but you may use base R too. patchwork combines separate ggplots into the same graphic with easy access to control layouts. limma is a Bioconductor package to analyze microarray data. It is not essential for our analysis, but may provide a more efficient implementation of the Wilcoxon rank sum test in differential expression analysis.


```{r library, echo=TRUE, warning=TRUE}
library(tidyverse)
library(Seurat)
library(dplyr) # data manipulation
library(ggplot2)
library(patchwork)
library(clustree)
library(SingleCellExperiment)
library(SingleR)
library(celldex)
library(scRNAseq)
library(pheatmap)
library(scran)
library(viridis)
library(ggforce)
library(gghalves)
library(ggridges)
library(scDblFinder)
library(cluster)
library(bluster)
```

##2 Load Data
#2.1 Load 12 10X Genomics Visium dataset into Seurat
A spatial gene expression dataset of mouse distal colon and tumor tissue collected by Space Ranger 2.0.0. will
be analyzed throughout the tutorial. Both the gene expression matrix and spatial imaging data are
necessary for the computational analysis.
The data files we will be using today include:
- a (filtered) feature / cell matrix HDF5 file (.h5)
- a spatial imaging data folder (.tar.gz)

```{r load, echo=TRUE, warning=TRUE}
E296_data <- Seurat::Load10X_Spatial(
  data.dir = "/stor/work/Fleet/BCG/2024_04.Fleet_visium_customgenome/02.count/E296_C-A1/outs/",
  filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "CR_1", filter.matrix = TRUE, to.upper = FALSE)
C507_data <- Seurat::Load10X_Spatial(
  data.dir = "/stor/work/Fleet/BCG/2024_04.Fleet_visium_customgenome/02.count/C507_A-A1/outs/",
  filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "ACR_1", filter.matrix = TRUE, to.upper = FALSE)
C569_data <- Seurat::Load10X_Spatial(
  data.dir = "/stor/work/Fleet/BCG/2024_04.Fleet_visium_customgenome/02.count/C569_C-B1/outs/",
  filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "ACR_2", filter.matrix = TRUE, to.upper = FALSE)
C822_data <- Seurat::Load10X_Spatial(
  data.dir = "/stor/work/Fleet/BCG/2024_04.Fleet_visium_customgenome/02.count/C822_C-C1/outs/",
  filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "AKCR_1", filter.matrix = TRUE, to.upper = FALSE)
C498_data <- Seurat::Load10X_Spatial(
  data.dir = "/stor/work/Fleet/BCG/2024_04.Fleet_visium_customgenome/02.count/C498_A-B1/outs/",
  filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "AKCR_2", filter.matrix = TRUE, to.upper = FALSE)
C531_data <- Seurat::Load10X_Spatial(
  data.dir = "/stor/work/Fleet/BCG/2024_04.Fleet_visium_customgenome/02.count/C531_A-D1/outs/",
  filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "AKCR_3", filter.matrix = TRUE, to.upper = FALSE)
C583_data <- Seurat::Load10X_Spatial(
  data.dir = "/stor/work/Fleet/BCG/2024_04.Fleet_visium_customgenome/02.count/C583_A-C1/outs/",
  filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "AKCR_4", filter.matrix = TRUE, to.upper = FALSE)
C520_data <- Seurat::Load10X_Spatial(
  data.dir = "/stor/work/Fleet/BCG/2024_04.Fleet_visium_customgenome/02.count/C520_B-A1/outs/",
  filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "AKCR_5", filter.matrix = TRUE, to.upper = FALSE)
C522_data <- Seurat::Load10X_Spatial(
  data.dir = "/stor/work/Fleet/BCG/2024_04.Fleet_visium_customgenome/02.count/C522_B-B1/outs/",
  filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "AKCR_6", filter.matrix = TRUE, to.upper = FALSE)
C565_data <- Seurat::Load10X_Spatial(
  data.dir = "/stor/work/Fleet/BCG/2024_04.Fleet_visium_customgenome/02.count/C565_C-D1/outs/",
  filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "ACR_TU1", filter.matrix = TRUE, to.upper = FALSE)
C515_data <- Seurat::Load10X_Spatial(
  data.dir = "/stor/work/Fleet/BCG/2024_04.Fleet_visium_customgenome/02.count/C515_B-C1/outs/",
  filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "ACR_TU2", filter.matrix = TRUE, to.upper = FALSE)
C533_data <- Seurat::Load10X_Spatial(
  data.dir = "/stor/work/Fleet/BCG/2024_04.Fleet_visium_customgenome/02.count/C533_B-D1/outs/",
  filename = "filtered_feature_bc_matrix.h5", assay = "Spatial", slice = "AKCR_TU1", filter.matrix = TRUE, to.upper = FALSE)
```

#2.2 View each dataset

```{r dataview, echo=TRUE, warning=TRUE}
E296_data
C507_data
C569_data
C822_data
C498_data
C583_data
C531_data
C520_data
C522_data
C565_data
C515_data
C533_data
```
## 3 Preprocess data
# 3.1. Quality control
A few common QC metrics include

- The number of unique genes detected in each sample (nFeature_Spatial).
- The total number of molecules detected within a sample (nCount_Spatial).
- The percentage of reads that map to the mitochondrial genome (percent.mt).

```{r qcmetrics, echo=TRUE, warning=TRUE}
E296_data[["percent.mt"]] <- PercentageFeatureSet(E296_data, pattern = "^mt-")
VlnPlot(
  E296_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt"), 
  pt.size = 0.1, ncol = 3) & 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
C507_data[["percent.mt"]] <- PercentageFeatureSet(C507_data, pattern = "^mt-")
VlnPlot(
  C507_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt"), 
  pt.size = 0.1, ncol = 3) & 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
C569_data[["percent.mt"]] <- PercentageFeatureSet(C569_data, pattern = "^mt-")
VlnPlot(
  C569_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt"), 
  pt.size = 0.1, ncol = 3) & 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
C822_data[["percent.mt"]] <- PercentageFeatureSet(C822_data, pattern = "^mt-")
VlnPlot(
  C822_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt"), 
  pt.size = 0.1, ncol = 3) & 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
C498_data[["percent.mt"]] <- PercentageFeatureSet(C498_data, pattern = "^mt-")
VlnPlot(
  C498_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt"), 
  pt.size = 0.1, ncol = 3) & 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
C531_data[["percent.mt"]] <- PercentageFeatureSet(C531_data, pattern = "^mt-")
VlnPlot(
  C531_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt"), 
  pt.size = 0.1, ncol = 3) & 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
C583_data[["percent.mt"]] <- PercentageFeatureSet(C583_data, pattern = "^mt-")
VlnPlot(
  C583_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt"), 
  pt.size = 0.1, ncol = 3) & 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
C520_data[["percent.mt"]] <- PercentageFeatureSet(C520_data, pattern = "^mt-")
VlnPlot(
  C520_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt"), 
  pt.size = 0.1, ncol = 3) & 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
C522_data[["percent.mt"]] <- PercentageFeatureSet(C522_data, pattern = "^mt-")
VlnPlot(
  C522_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt"), 
  pt.size = 0.1, ncol = 3) & 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
C565_data[["percent.mt"]] <- PercentageFeatureSet(C565_data, pattern = "^mt-")
VlnPlot(
  C565_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt"), 
  pt.size = 0.1, ncol = 3) & 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
C515_data[["percent.mt"]] <- PercentageFeatureSet(C515_data, pattern = "^mt-")
VlnPlot(
  C515_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt"), 
  pt.size = 0.1, ncol = 3) & 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
C533_data[["percent.mt"]] <- PercentageFeatureSet(C533_data, pattern = "^mt-")
VlnPlot(
  C533_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt"), 
  pt.size = 0.1, ncol = 3) & 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
```
#3.2 Subsetting samples based on observation

After observing these plots from each sample, I decided to filter out molecules < 1500 & a percentage of reads that map to the mitochondrial genome < 20%.


```{r subsetting, echo=TRUE, warning=TRUE}
E296_subset <- subset(E296_data, subset = nCount_Spatial > 1500 & percent.mt < 20)
print(paste("Filter out", ncol(E296_data) - ncol(E296_subset), 
            "samples because of the outlier QC metrics, with", ncol(E296_subset), "samples left."))
C507_subset <- subset(C507_data, subset = nCount_Spatial > 1500 & percent.mt < 20)
print(paste("Filter out", ncol(C507_data) - ncol(C507_subset), 
            "samples because of the outlier QC metrics, with", ncol(C507_subset), "samples left."))
C569_subset <- subset(C569_data, subset = nCount_Spatial > 1500 & percent.mt < 20)
print(paste("Filter out", ncol(C569_data) - ncol(C569_subset), 
            "samples because of the outlier QC metrics, with", ncol(C569_subset), "samples left."))
C822_subset <- subset(C822_data, subset = nCount_Spatial > 1500 & percent.mt < 20)
print(paste("Filter out", ncol(C822_data) - ncol(C822_subset), 
            "samples because of the outlier QC metrics, with", ncol(C822_subset), "samples left."))
C498_subset <- subset(C498_data, subset = nCount_Spatial > 1500 & percent.mt < 20)
print(paste("Filter out", ncol(C498_data) - ncol(C498_subset), 
            "samples because of the outlier QC metrics, with", ncol(C498_subset), "samples left."))
C531_subset <- subset(C531_data, subset = nCount_Spatial > 1500 & percent.mt < 20)
print(paste("Filter out", ncol(C531_data) - ncol(C531_subset), 
            "samples because of the outlier QC metrics, with", ncol(C531_subset), "samples left."))
C583_subset <- subset(C583_data, subset = nCount_Spatial > 1500 & percent.mt < 20)
print(paste("Filter out", ncol(C583_data) - ncol(C583_subset), 
            "samples because of the outlier QC metrics, with", ncol(C583_subset), "samples left."))
C520_subset <- subset(C520_data, subset = nCount_Spatial > 1500 & percent.mt < 20)
print(paste("Filter out", ncol(C520_data) - ncol(C520_subset), 
            "samples because of the outlier QC metrics, with", ncol(C520_subset), "samples left."))
C522_subset <- subset(C522_data, subset = nCount_Spatial > 1500 & percent.mt < 20)
print(paste("Filter out", ncol(C522_data) - ncol(C522_subset), 
            "samples because of the outlier QC metrics, with", ncol(C522_subset), "samples left."))
C565_subset <- subset(C565_data, subset = nCount_Spatial > 1500 & percent.mt < 20)
print(paste("Filter out", ncol(C565_data) - ncol(C565_subset), 
            "samples because of the outlier QC metrics, with", ncol(C565_subset), "samples left."))
C515_subset <- subset(C515_data, subset = nCount_Spatial > 1500 & percent.mt < 20)
print(paste("Filter out", ncol(C515_data) - ncol(C515_subset), 
            "samples because of the outlier QC metrics, with", ncol(C515_subset), "samples left."))
C533_subset <- subset(C533_data, subset = nCount_Spatial > 1500 & percent.mt < 20)
print(paste("Filter out", ncol(C533_data) - ncol(C533_subset), 
            "samples because of the outlier QC metrics, with", ncol(C533_subset), "samples left."))
```
# 3.3. Gene Expression Visualization

The SpatialFeaturePlot() function in Seurat extends FeaturePlot(), and can overlay molecular data on top of tissue histology. I used this feature to see the subsetted out spots overlayed on tissue image.

```{r postqcview,fig.height=8, echo=TRUE, warning=TRUE}
plot <- SpatialFeaturePlot(
  E296_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "right") & ggtitle("E296_subset")
plot
png(filename = "../05.Analysis_in_R/Preprocessing/E296_subset.png", height = 2000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialFeaturePlot(
  C507_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "right") & ggtitle("C507_subset")
plot
png(filename = "../05.Analysis_in_R/Preprocessing/C507_subset.png", height = 2000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialFeaturePlot(
  C569_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "right") & ggtitle("C569_subset")
plot
png(filename = "../05.Analysis_in_R/Preprocessing/C569_subset.png", height = 2000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialFeaturePlot(
  C822_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "right") & ggtitle("C822_subset")
plot
png(filename = "../05.Analysis_in_R/Preprocessing/C822_subset.png", height = 2000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialFeaturePlot(
  C498_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "right") & ggtitle("C498_subset")
plot
png(filename = "../05.Analysis_in_R/Preprocessing/C498_subset.png", height = 2000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialFeaturePlot(
  C531_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "right") & ggtitle("C531_subset")
plot
png(filename = "../05.Analysis_in_R/Preprocessing/C531_subset.png", height = 2000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialFeaturePlot(
  C583_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "right") & ggtitle("C583_subset")
plot
png(filename = "../05.Analysis_in_R/Preprocessing/C583_subset.png", height = 2000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialFeaturePlot(
  C520_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "right") & ggtitle("C520_subset")
plot
png(filename = "../05.Analysis_in_R/Preprocessing/C520_subset.png", height = 2000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialFeaturePlot(
  C522_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "right") & ggtitle("C522_subset")
plot
png(filename = "../05.Analysis_in_R/Preprocessing/C522_subset.png", height = 2000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialFeaturePlot(
  C565_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "right") & ggtitle("C565_subset")
plot
png(filename = "../05.Analysis_in_R/Preprocessing/C565_subset.png", height = 2000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialFeaturePlot(
  C515_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "right") & ggtitle("C515_subset")
plot
png(filename = "../05.Analysis_in_R/Preprocessing/C515_subset.png", height = 2000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialFeaturePlot(
  C533_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "right") & ggtitle("C533_subset")
plot
png(filename = "../05.Analysis_in_R/Preprocessing/C533_subset.png", height = 2000, width = 6000, res = 300)
print(plot)
dev.off()
```

Filtering successfully discarded spots that appeared as low quality.

# 3.4 Normalization and Re-labeling

The variance of molecular counts expresses spatial heterogeneity which cannot be solely explained by technical noise. Satija Lab and Collaborators recommends normalization using SCTransform (Hafemeister and Satija, 2019) in order to account for technical bias while preserving true biological differences. We will also change the "orig.ident" of each sample to unique sample number (genotype+number, e.g. AKCR_3). By changing their orig.ident to unique names, we can still call up a single sample when they are pooled.

```{r norm, echo=TRUE, warning=TRUE}
E296_norm <- SCTransform(E296_subset, assay = "Spatial", verbose = FALSE)
C507_norm <- SCTransform(C507_subset, assay = "Spatial", verbose = FALSE)
C569_norm <- SCTransform(C569_subset, assay = "Spatial", verbose = FALSE)
C822_norm <- SCTransform(C822_subset, assay = "Spatial", verbose = FALSE)
C498_norm <- SCTransform(C498_subset, assay = "Spatial", verbose = FALSE)
C531_norm <- SCTransform(C531_subset, assay = "Spatial", verbose = FALSE)
C583_norm <- SCTransform(C583_subset, assay = "Spatial", verbose = FALSE)
C520_norm <- SCTransform(C520_subset, assay = "Spatial", verbose = FALSE)
C522_norm <- SCTransform(C522_subset, assay = "Spatial", verbose = FALSE)
C565_norm <- SCTransform(C565_subset, assay = "Spatial", verbose = FALSE)
C515_norm <- SCTransform(C515_subset, assay = "Spatial", verbose = FALSE)
C533_norm <- SCTransform(C533_subset, assay = "Spatial", verbose = FALSE)
```

```{r changeorigident, echo=TRUE, warning=TRUE}
E296_subset$orig.ident <- "CR_1"
C507_subset$orig.ident <- "ACR_1"
C569_subset$orig.ident <- "ACR_2"
C822_subset$orig.ident <- "AKCR_1"
C498_subset$orig.ident <- "AKCR_2"
C531_subset$orig.ident <- "AKCR_3"
C583_subset$orig.ident <- "AKCR_4"
C520_subset$orig.ident <- "AKCR_5"
C522_subset$orig.ident <- "AKCR_6"
C565_subset$orig.ident <- "ACR_TU1"
C515_subset$orig.ident <- "ACR_TU2"
C533_subset$orig.ident <- "AKCR_TU1"
E296_norm$orig.ident <- "CR_1"
C507_norm$orig.ident <- "ACR_1"
C569_norm$orig.ident <- "ACR_2"
C822_norm$orig.ident <- "AKCR_1"
C498_norm$orig.ident <- "AKCR_2"
C531_norm$orig.ident <- "AKCR_3"
C583_norm$orig.ident <- "AKCR_4"
C520_norm$orig.ident <- "AKCR_5"
C522_norm$orig.ident <- "AKCR_6"
C565_norm$orig.ident <- "ACR_TU1"
C515_norm$orig.ident <- "ACR_TU2"
C533_norm$orig.ident <- "AKCR_TU1"
```

#3.5 View highly variable genes 

```{r variablefeaturespersample, echo=TRUE, warning=TRUE}
top20 <- head(VariableFeatures(E296_norm), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(E296_norm)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/E296variablegene.png", height = 1500, width = 2500, res = 300)
print(plot2)
dev.off()

top20 <- head(VariableFeatures(C507_norm), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(C507_norm)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/C507variablegene.png", height = 1500, width = 2500, res = 300)
print(plot2)
dev.off()

top20 <- head(VariableFeatures(C569_norm), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(C569_norm)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/C569variablegene.png", height = 1500, width = 2500, res = 300)
print(plot2)
dev.off()

top20 <- head(VariableFeatures(C498_norm), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(C498_norm)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/C498variablegene.png", height = 1500, width = 2500, res = 300)
print(plot2)
dev.off()

top20 <- head(VariableFeatures(C822_norm), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(C822_norm)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/C822variablegene.png", height = 1500, width = 2500, res = 300)
print(plot2)
dev.off()

top20 <- head(VariableFeatures(C583_norm), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(C583_norm)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/C583variablegene.png", height = 1500, width = 2500, res = 300)
print(plot2)
dev.off()

top20 <- head(VariableFeatures(C531_norm), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(C531_norm)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/C531variablegene.png", height = 1500, width = 2500, res = 300)
print(plot2)
dev.off()

top20 <- head(VariableFeatures(C520_norm), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(C520_norm)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/C520variablegene.png", height = 1500, width = 2500, res = 300)
print(plot2)
dev.off()

top20 <- head(VariableFeatures(C522_norm), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(C522_norm)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/C522variablegene.png", height = 1500, width = 2500, res = 300)
print(plot2)
dev.off()

top20 <- head(VariableFeatures(C515_norm), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(C515_norm)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/C515variablegene.png", height = 1500, width = 2500, res = 300)
print(plot2)
dev.off()

top20 <- head(VariableFeatures(C565_norm), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(C565_norm)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/C565variablegene.png", height = 1500, width = 2500, res = 300)
print(plot2)
dev.off()

top20 <- head(VariableFeatures(C533_norm), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(C533_norm)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/C533variablegene.png", height = 1500, width = 2500, res = 300)
print(plot2)
dev.off()
```
```{r poolspatialsamples, echo=TRUE, warning=TRUE}
spatial_pooled <- merge(x = E296_subset, y = c(C507_subset, C569_subset, C822_subset, 
                                               C498_subset, C531_subset, C583_subset, 
                                               C520_subset, C522_subset, C565_subset, 
                                               C515_subset, C533_subset))
order <- c("CR_1","ACR_1","ACR_2","AKCR_1","AKCR_2","AKCR_3","AKCR_4","AKCR_5", "AKCR_6", "ACR_TU1", "ACR_TU2", "AKCR_TU1")
spatial_pooled[["orig.ident"]] <- factor(x = spatial_pooled@meta.data$orig.ident, levels = order)
p<- VlnPlot(spatial_pooled, group.by = c("orig.ident"), features = c("nCount_Spatial", "nFeature_Spatial"), pt.size = 0.1, ncol = 1)
p
png(filename = "../05.Analysis_in_R/3.Clustering/presctransform.png", height = 2350, width = 1500, res = 300)
print(p)
dev.off()
```

```{r join layer, echo=TRUE, warning=TRUE}
spatial_pooled
spatial_joined <- JoinLayers(spatial_pooled)
spatial_joined
```

```{r percellbysample_presctransform, echo=TRUE, warning=TRUE}
temp_labels <- spatial_joined@meta.data %>%
group_by(orig.ident) %>%
tally()
p1 <- ggplot() +
geom_half_violin(
data = spatial_joined@meta.data, aes(orig.ident, nCount_Spatial, fill = orig.ident),
side = 'l', show.legend = FALSE, trim = FALSE
) +
geom_half_boxplot(
data = spatial_joined@meta.data, aes(orig.ident, nCount_Spatial, fill = orig.ident),
side = 'r', outlier.color = NA, width = 0.4, show.legend = FALSE
) +
geom_text(
data = temp_labels,
aes(x = orig.ident, y = -Inf, label = paste0('n = ', format(n, big.mark = ',', trim = TRUE)), vjust = -1),
color = 'black', size = 2.8
) +
scale_y_continuous(labels = scales::comma, expand = c(0.08,0)) +
theme_bw() +
labs(x = '', y = 'Number of transcripts') +
theme(
panel.grid.major.x = element_blank(),
axis.title.x = element_blank()
)
p2 <- ggplot() +
geom_half_violin(
data = spatial_joined@meta.data, aes(orig.ident, nFeature_Spatial, fill = orig.ident),
side = 'l', show.legend = FALSE, trim = FALSE
) +
geom_half_boxplot(
data = spatial_joined@meta.data, aes(orig.ident, nFeature_Spatial, fill = orig.ident),
side = 'r', outlier.color = NA, width = 0.4, show.legend = FALSE
) +
geom_text(
data = temp_labels,
aes(x = orig.ident, y = -Inf, label = paste0('n = ', format(n, big.mark = ',', trim = TRUE)), vjust = -1),
color = 'black', size = 2.8
) +
scale_y_continuous(
name = 'Number of expressed genes',
labels = scales::comma, expand = c(0.08,0)
) +
theme_bw() +
theme(
panel.grid.major.x = element_blank(),
axis.title.x = element_blank()
)
p1
p2

```

```{r integratedvariablefeatureplot3, echo=TRUE, warning=TRUE}
spatial_obj <- SCTransform(spatial_joined, assay = "Spatial", ncells = 14799, verbose = TRUE)

p<- VlnPlot(spatial_obj, group.by = c("orig.ident"), features = c("nCount_SCT", "nFeature_SCT"), pt.size = 0.1, ncol = 1)
p
png(filename = "../05.Analysis_in_R/3.Clustering/postsctransform.png", height = 2350, width = 1500, res = 300)
print(p)
dev.off()

# Merge normalized samples
DefaultAssay(spatial_obj) <- "SCT"

```


```{r percellbysample_postsctransform, echo=TRUE, warning=TRUE}
temp_labels <- spatial_obj@meta.data %>%
group_by(orig.ident) %>%
tally()
p1 <- ggplot() +
geom_half_violin(
data = spatial_obj@meta.data, aes(orig.ident, nCount_SCT, fill = orig.ident),
side = 'l', show.legend = FALSE, trim = FALSE
) +
geom_half_boxplot(
data = spatial_obj@meta.data, aes(orig.ident, nCount_SCT, fill = orig.ident),
side = 'r', outlier.color = NA, width = 0.4, show.legend = FALSE
) +
geom_text(
data = temp_labels,
aes(x = orig.ident, y = -Inf, label = paste0('n = ', format(n, big.mark = ',', trim = TRUE)), vjust = -1),
color = 'black', size = 2.8
) +
scale_y_continuous(labels = scales::comma, expand = c(0.08,0)) +
theme_bw() +
labs(x = '', y = 'Number of transcripts') +
theme(
panel.grid.major.x = element_blank(),
axis.title.x = element_blank()
)
p2 <- ggplot() +
geom_half_violin(
data = spatial_obj@meta.data, aes(orig.ident, nFeature_SCT, fill = orig.ident),
side = 'l', show.legend = FALSE, trim = FALSE
) +
geom_half_boxplot(
data = spatial_obj@meta.data, aes(orig.ident, nFeature_SCT, fill = orig.ident),
side = 'r', outlier.color = NA, width = 0.4, show.legend = FALSE
) +
geom_text(
data = temp_labels,
aes(x = orig.ident, y = -Inf, label = paste0('n = ', format(n, big.mark = ',', trim = TRUE)), vjust = -1),
color = 'black', size = 2.8
) +
scale_y_continuous(
name = 'Number of expressed genes',
labels = scales::comma, expand = c(0.08,0)
) +
theme_bw() +
theme(
panel.grid.major.x = element_blank(),
axis.title.x = element_blank()
)
p1
p2
```


## Using 2000 variable genes

```{r 2000genes, echo=TRUE, warning=TRUE}
spatial_2000 <- SCTransform(spatial_joined, assay = "Spatial", ncells = 14799, residual.features = NULL, variable.features.n = 2000,  verbose = TRUE)
```

```{r 2000genes2, echo=TRUE, warning=TRUE}
top30 <- head(VariableFeatures(spatial_2000), 30)
plot1 <- VariableFeaturePlot(spatial_2000)
plot2 <- LabelPoints(plot = plot1, points = top30, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/2000variablegene.png", height = 2000, width = 2500, res = 300)
print(plot2)
dev.off()
```


```{r 2000dimentionality reduction, echo=TRUE, warning=TRUE}
spatial_2000 <- RunPCA(spatial_2000, verbose = FALSE)
#Principal component loadings for top genes for the first 2 principal components. The loading can be interpreted as the weights for each original variable when calculating the principal components.
VizDimLoadings(spatial_2000, dims = 1:2, reduction = "pca")
#Heatmaps of the top 20 PCs. Cells and features are ordered according to their PCA scores and the 500 most extreme cells are selected to be shown in these plots.
DimHeatmap(spatial_2000, dims = 1:20, cells = 500, balanced = TRUE)
#Elbow plot for top principal components. Generally, PCs are used for further analysis which capture the majority of the variation seen in the data (before the elbow which is seen once the standard deviation associated with the PCs becomes small and does not change much between PCx and PCx+1.
ElbowPlot(spatial_2000, ndims = 50)
```

Based on this plot, we could roughly determine the majority of the variation by where the elbow occurs (touches the ground) to be between PC17-PC20. Determine percent of variation associated with each PC. While this gives us a good rough idea of the number of PCs needed to be included, a more quantitative approach may be a bit more reliable. We can calculate where the principal components start to elbow by taking the larger value of:

The point where the principal components only contribute 5% of standard deviation and the principal components cumulatively contribute 90% of the standard deviation.
The point where the percent change in variation between the consecutive PCs is less than 0.1%.
We will start by calculating the first metric:

```{r 2000firstmetric, echo=TRUE, warning=TRUE}
pct <- spatial_2000[["pca"]]@stdev / sum(spatial_2000[["pca"]]@stdev) * 100
# Calculate cumulative percents for each PC
cumu <- cumsum(pct)
# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
co1 <- which(cumu > 90 & pct < 5)[1]
co1
```
The first metric returns PC41 as the PC matching these requirements. Let’s check the second metric, which identifies the PC where the percent change in variation between consecutive PCs is less than 0.1%:

```{r 2000secondmetric, echo=TRUE, warning=TRUE}
# Determine the difference between variation of PC and subsequent PC
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1
# last point where change of % of variation is more than 0.1%.
co2
```
This second metric returns PC20. Usually, we would choose the minimum of these two metrics as the PCs covering the majority of the variation in the data.

```{r 2000minmetric, echo=TRUE, warning=TRUE}
# Minimum of the two calculation
pcs <- min(co1, co2)
pcs
```

```{r 2000clustering, echo=TRUE, warning=TRUE}
spatial_2000 <- FindNeighbors(spatial_2000, dims = 1:20)
spatial_2000 <- FindClusters(spatial_2000, verbose = FALSE)
spatial_2000 <- RunUMAP(spatial_2000, dims = 1:20)
```
```{r 2000multiresolution, echo=TRUE, warning=TRUE}
resolution_values <- c(0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3)
spatial_2000 <- FindClusters(spatial_2000, assay = "SCT", resolution = resolution_values)
plot <- clustree(spatial_2000, prefix = "SCT_snn_res.")
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/clustree.png", height = 3000, width = 3000, res = 300)
print(plot)
dev.off()
```

```{r 2000umapindividualsamplessave, echo=TRUE, warning=TRUE}
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = "SCT_snn_res.0.4", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3)
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/umappersample0.4.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = "SCT_snn_res.0.5", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3)
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/umappersample0.5.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = "SCT_snn_res.0.6", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/umappersample0.6.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = "SCT_snn_res.0.7", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/umappersample0.7.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = "SCT_snn_res.0.8", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/umappersample0.8.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = "SCT_snn_res.0.9", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/umappersample0.9.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
```
```{r 2000spatialumapindividualsamplessave, echo=TRUE, warning=TRUE}
plot <- SpatialDimPlot(spatial_2000, group.by = "SCT_snn_res.0.4", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/spatialumapres0.4.png", height = 8000, width = 6000, res = 300)
plot
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_2000, group.by = "SCT_snn_res.0.5", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/spatialumapres0.5.png", height = 8000, width = 6000, res = 300)
plot
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_2000, group.by = "SCT_snn_res.0.6", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/spatialumapres0.6.png", height = 8000, width = 6000, res = 300)
plot
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_2000, group.by = "SCT_snn_res.0.7", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/spatialumapres0.7.png", height = 8000, width = 6000, res = 300)
plot
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_2000, group.by = "SCT_snn_res.0.8", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/spatialumapres0.8.png", height = 8000, width = 6000, res = 300)
plot
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_2000, group.by = "SCT_snn_res.0.9", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/spatialumapres0.9.png", height = 8000, width = 6000, res = 300)
plot
print(plot)
dev.off()
```

```{r 2000pooledplot2, echo=TRUE, warning=TRUE}
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = c("SCT_snn_res.0.4", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/umap_res_0.4.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = c("SCT_snn_res.0.5", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/umap_res_0.5.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = c("SCT_snn_res.0.6", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/umap_res_0.6.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = c("SCT_snn_res.0.7", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/umap_res_0.7.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = c("SCT_snn_res.0.8", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/umap_res_0.8.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = c("SCT_snn_res.0.9", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/umap_res_0.9.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
```

```{r 2000pooledplot3, echo=TRUE, warning=TRUE}
plot <- DimPlot(spatial_2000, reduction = "umap", group.by = c("SCT_snn_res.0.4", "SCT_snn_res.0.5", "SCT_snn_res.0.6", "SCT_snn_res.0.7", "SCT_snn_res.0.8", "SCT_snn_res.0.9"), label = TRUE, label.size = 10, pt.size = 0.5, ncol = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/0.4to0.9_umap.png", height = 6000, width = 10000, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_2000, reduction = "pca", group.by = c("SCT_snn_res.0.4", "SCT_snn_res.0.5", "SCT_snn_res.0.6", "SCT_snn_res.0.7", "SCT_snn_res.0.8", "SCT_snn_res.0.9"), label = TRUE, label.size = 10, pt.size = 0.5, ncol = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/2000 genes 20 PCs used/0.4to0.9_pca.png", height = 6000, width = 10000, res = 300)
print(plot)
dev.off()
```

## Using 3000 variable genes

```{r 3000genes, echo=TRUE, warning=TRUE}
spatial_3000 <- SCTransform(spatial_joined, assay = "Spatial", ncells = 14799, residual.features = NULL, variable.features.n = 3000,  verbose = TRUE)
```

```{r 3000genes2, echo=TRUE, warning=TRUE}
top30 <- head(VariableFeatures(spatial_3000), 30)
plot1 <- VariableFeaturePlot(spatial_3000)
plot2 <- LabelPoints(plot = plot1, points = top30, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/3000variablegene.png", height = 2000, width = 2500, res = 300)
print(plot2)
dev.off()
```

# Determine the ‘dimensionality’ of the dataset

```{r 3000dimentionality reduction, echo=TRUE, warning=TRUE}
spatial_3000 <- RunPCA(spatial_3000, verbose = FALSE)
#Principal component loadings for top genes for the first 2 principal components. The loading can be interpreted as the weights for each original variable when calculating the principal components.
VizDimLoadings(spatial_3000, dims = 1:2, reduction = "pca")
#Heatmaps of the top 20 PCs. Cells and features are ordered according to their PCA scores and the 500 most extreme cells are selected to be shown in these plots.
DimHeatmap(spatial_3000, dims = 1:20, cells = 500, balanced = TRUE)
#Elbow plot for top principal components. Generally, PCs are used for further analysis which capture the majority of the variation seen in the data (before the elbow which is seen once the standard deviation associated with the PCs becomes small and does not change much between PCx and PCx+1.
ElbowPlot(spatial_3000, ndims = 40)
```

Based on this plot, we could roughly determine the majority of the variation by where the elbow occurs (touches the ground) to be between PC17-PC20. Determine percent of variation associated with each PC. While this gives us a good rough idea of the number of PCs needed to be included, a more quantitative approach may be a bit more reliable. We can calculate where the principal components start to elbow by taking the larger value of:

The point where the principal components only contribute 5% of standard deviation and the principal components cumulatively contribute 90% of the standard deviation.
The point where the percent change in variation between the consecutive PCs is less than 0.1%.
We will start by calculating the first metric:

```{r 3000firstmetric, echo=TRUE, warning=TRUE}
pct <- spatial_3000[["pca"]]@stdev / sum(spatial_3000[["pca"]]@stdev) * 100
# Calculate cumulative percents for each PC
cumu <- cumsum(pct)
# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
co1 <- which(cumu > 90 & pct < 5)[1]
co1
```
The first metric returns PC41 as the PC matching these requirements. Let’s check the second metric, which identifies the PC where the percent change in variation between consecutive PCs is less than 0.1%:

```{r 3000secondmetric, echo=TRUE, warning=TRUE}
# Determine the difference between variation of PC and subsequent PC
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1
# last point where change of % of variation is more than 0.1%.
co2
```
This second metric returns PC20. Usually, we would choose the minimum of these two metrics as the PCs covering the majority of the variation in the data.

```{r 3000minmetric, echo=TRUE, warning=TRUE}
# Minimum of the two calculation
pcs <- min(co1, co2)
pcs
```



```{r 3000clustering, echo=TRUE, warning=TRUE}
spatial_3000 <- FindNeighbors(spatial_3000, dims = 1:20)
spatial_3000 <- FindClusters(spatial_3000, verbose = FALSE)
spatial_3000 <- RunUMAP(spatial_3000, dims = 1:20)
```
```{r 3000multiresolution, echo=TRUE, warning=TRUE}
resolution_values <- c(0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3)
spatial_3000 <- FindClusters(spatial_3000, assay = "SCT", resolution = resolution_values)
plot <- clustree(spatial_3000, prefix = "SCT_snn_res.")
plot
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/clustree.png", height = 3000, width = 3000, res = 300)
print(plot)
dev.off()
```

```{r 3000umapindividualsamplessave, echo=TRUE, warning=TRUE}
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = "SCT_snn_res.0.4", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/umappersample0.4.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = "SCT_snn_res.0.5", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/umappersample0.5.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = "SCT_snn_res.0.6", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/umappersample0.6.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = "SCT_snn_res.0.7", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/umappersample0.7.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = "SCT_snn_res.0.8", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/umappersample0.8.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = "SCT_snn_res.0.9", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/umappersample0.9.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
```
```{r 3000spatialumapindividualsamplessave, echo=TRUE, warning=TRUE}
plot <- SpatialDimPlot(spatial_3000, group.by = "SCT_snn_res.0.4", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/spatialumapres0.4.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_3000, group.by = "SCT_snn_res.0.5", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/spatialumapres0.5.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_3000, group.by = "SCT_snn_res.0.6", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/spatialumapres0.6.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_3000, group.by = "SCT_snn_res.0.7", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/spatialumapres0.7.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_3000, group.by = "SCT_snn_res.0.8", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/spatialumapres0.8.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_3000, group.by = "SCT_snn_res.0.9", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/spatialumapres0.9.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
```
```{r 3000pooledplot2, echo=TRUE, warning=TRUE}
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = c("SCT_snn_res.0.4", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/umap_res_0.4.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = c("SCT_snn_res.0.5", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/umap_res_0.5.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = c("SCT_snn_res.0.6", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/umap_res_0.6.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = c("SCT_snn_res.0.7", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/umap_res_0.7.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = c("SCT_snn_res.0.8", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/umap_res_0.8.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = c("SCT_snn_res.0.9", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/umap_res_0.9.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
```

```{r 3000pooledplot3, echo=TRUE, warning=TRUE}
plot <- DimPlot(spatial_3000, reduction = "umap", group.by = c("SCT_snn_res.0.4", "SCT_snn_res.0.5", "SCT_snn_res.0.6", "SCT_snn_res.0.7", "SCT_snn_res.0.8", "SCT_snn_res.0.9"), label = TRUE, label.size = 10, pt.size = 0.5, ncol = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/0.4to0.9_umap.png", height = 6000, width = 10000, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_3000, reduction = "pca", group.by = c("SCT_snn_res.0.4", "SCT_snn_res.0.5", "SCT_snn_res.0.6", "SCT_snn_res.0.7", "SCT_snn_res.0.8", "SCT_snn_res.0.9"), label = TRUE, label.size = 10, pt.size = 0.5, ncol = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/3000 genes 20 PCs used/0.4to0.9_pca.png", height = 6000, width = 10000, res = 300)
print(plot)
dev.off()
```

## Using 4000 variable genes 

```{r 4000genes, echo=TRUE, warning=TRUE}

spatial_4000 <- SCTransform(spatial_joined, assay = "Spatial", ncells = 14799, residual.features = NULL, variable.features.n = 4000,  verbose = TRUE)
```

```{r 4000genes2, echo=TRUE, warning=TRUE}
top30 <- head(VariableFeatures(spatial_4000), 30)
plot1 <- VariableFeaturePlot(spatial_4000)
plot2 <- LabelPoints(plot = plot1, points = top30, repel = TRUE)
plot2
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/4000variablegene.png", height = 2000, width = 2500, res = 300)
print(plot2)
dev.off()
```


```{r 4000genes dimentionality reduction, echo=TRUE, warning=TRUE}
spatial_4000 <- RunPCA(spatial_4000, verbose = FALSE)
#Principal component loadings for top genes for the first 2 principal components. The loading can be interpreted as the weights for each original variable when calculating the principal components.
VizDimLoadings(spatial_4000, dims = 1:2, reduction = "pca")
#Heatmaps of the top 20 PCs. Cells and features are ordered according to their PCA scores and the 500 most extreme cells are selected to be shown in these plots.
DimHeatmap(spatial_4000, dims = 1:20, cells = 500, balanced = TRUE)
#Elbow plot for top principal components. Generally, PCs are used for further analysis which capture the majority of the variation seen in the data (before the elbow which is seen once the standard deviation associated with the PCs becomes small and does not change much between PCx and PCx+1.
ElbowPlot(spatial_4000, ndims = 40)
```

Based on this plot, we could roughly determine the majority of the variation by where the elbow occurs (touches the ground) to be between PC17-PC20. Determine percent of variation associated with each PC. While this gives us a good rough idea of the number of PCs needed to be included, a more quantitative approach may be a bit more reliable. We can calculate where the principal components start to elbow by taking the larger value of:

The point where the principal components only contribute 5% of standard deviation and the principal components cumulatively contribute 90% of the standard deviation.
The point where the percent change in variation between the consecutive PCs is less than 0.1%.
We will start by calculating the first metric:

```{r 4000genesfirstmetric, echo=TRUE, warning=TRUE}
pct <- spatial_4000[["pca"]]@stdev / sum(spatial_4000[["pca"]]@stdev) * 100
# Calculate cumulative percents for each PC
cumu <- cumsum(pct)
# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
co1 <- which(cumu > 90 & pct < 5)[1]
co1
```
The first metric returns PC41 as the PC matching these requirements. Let’s check the second metric, which identifies the PC where the percent change in variation between consecutive PCs is less than 0.1%:

```{r 4000genessecondmetric, echo=TRUE, warning=TRUE}
# Determine the difference between variation of PC and subsequent PC
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1
# last point where change of % of variation is more than 0.1%.
co2
```
This second metric returns PC18. Usually, we would choose the minimum of these two metrics as the PCs covering the majority of the variation in the data.

```{r 4000genesminmetric, echo=TRUE, warning=TRUE}
# Minimum of the two calculation
pcs <- min(co1, co2)
pcs
```

```{r 4000genesclustering, echo=TRUE, warning=TRUE}
spatial_4000 <- FindNeighbors(spatial_4000, dims = 1:20)
spatial_4000 <- FindClusters(spatial_4000, verbose = FALSE)
spatial_4000 <- RunUMAP(spatial_4000, dims = 1:20)
```
```{r 4000genesmultiresolution, echo=TRUE, warning=TRUE}
resolution_values <- c(0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3)
spatial_4000 <- FindClusters(spatial_4000, assay = "SCT", resolution = resolution_values)
plot <- clustree(spatial_4000, prefix = "SCT_snn_res.")
plot
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/clustree.png", height = 3000, width = 3000, res = 300)
print(plot)
dev.off()
```

```{r 4000genesumapindividualsamplessave, echo=TRUE, warning=TRUE}
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = "SCT_snn_res.0.4", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/umappersample0.4.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = "SCT_snn_res.0.5", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/umappersample0.5.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = "SCT_snn_res.0.6", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/umappersample0.6.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = "SCT_snn_res.0.7", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/umappersample0.7.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = "SCT_snn_res.0.8", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/umappersample0.8.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = "SCT_snn_res.0.9", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/umappersample0.9.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
```
```{r 4000genesspatialumapindividualsamplessave, echo=TRUE, warning=TRUE}
plot <- SpatialDimPlot(spatial_4000, group.by = "SCT_snn_res.0.4", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/spatialumapres0.4.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_4000, group.by = "SCT_snn_res.0.5", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/spatialumapres0.5.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_4000, group.by = "SCT_snn_res.0.6", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/spatialumapres0.6.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_4000, group.by = "SCT_snn_res.0.7", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/spatialumapres0.7.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_4000, group.by = "SCT_snn_res.0.8", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/spatialumapres0.8.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_4000, group.by = "SCT_snn_res.0.9", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/spatialumapres0.9.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
```

```{r 4000genespooledplot2, echo=TRUE, warning=TRUE}
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = c("SCT_snn_res.0.4", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/umap_res_0.4.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = c("SCT_snn_res.0.5", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/umap_res_0.5.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = c("SCT_snn_res.0.6", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/umap_res_0.6.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = c("SCT_snn_res.0.7", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/umap_res_0.7.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = c("SCT_snn_res.0.8", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/umap_res_0.8.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = c("SCT_snn_res.0.9", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/umap_res_0.9.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
```

```{r 4000genespooledplot3, echo=TRUE, warning=TRUE}
plot <- DimPlot(spatial_4000, reduction = "umap", group.by = c("SCT_snn_res.0.4", "SCT_snn_res.0.5", "SCT_snn_res.0.6", "SCT_snn_res.0.7", "SCT_snn_res.0.8", "SCT_snn_res.0.9"), label = TRUE, label.size = 10, pt.size = 0.5, ncol = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/0.4to0.9_umap.png", height = 6000, width = 10000, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_4000, reduction = "pca", group.by = c("SCT_snn_res.0.4", "SCT_snn_res.0.5", "SCT_snn_res.0.6", "SCT_snn_res.0.7", "SCT_snn_res.0.8", "SCT_snn_res.0.9"), label = TRUE, label.size = 10, pt.size = 0.5, ncol = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/4000 genes 20 PCs used/0.4to0.9_pca.png", height = 6000, width = 10000, res = 300)
print(plot)
dev.off()
```
## Using 6000 variable genes

```{r 6000genes, echo=TRUE, warning=TRUE}
spatial_6000 <- SCTransform(spatial_joined, assay = "Spatial", ncells = 14799, residual.features = NULL, variable.features.n = 6000,  verbose = TRUE)
```

```{r 6000genes2, echo=TRUE, warning=TRUE}
top30 <- head(VariableFeatures(spatial_6000), 30)
plot1 <- VariableFeaturePlot(spatial_6000)
plot2 <- LabelPoints(plot = plot1, points = top30, repel = TRUE)
  plot2
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/6000variablegene.png", height = 2000, width = 2500, res = 300)
print(plot2)
dev.off()
```


```{r 6000dimentionality reduction, echo=TRUE, warning=TRUE}
spatial_6000 <- RunPCA(spatial_6000, verbose = FALSE)
#Principal component loadings for top genes for the first 2 principal components. The loading can be interpreted as the weights for each original variable when calculating the principal components.
VizDimLoadings(spatial_6000, dims = 1:2, reduction = "pca")
#Heatmaps of the top 20 PCs. Cells and features are ordered according to their PCA scores and the 500 most extreme cells are selected to be shown in these plots.
DimHeatmap(spatial_6000, dims = 1:20, cells = 500, balanced = TRUE)
#Elbow plot for top principal components. Generally, PCs are used for further analysis which capture the majority of the variation seen in the data (before the elbow which is seen once the standard deviation associated with the PCs becomes small and does not change much between PCx and PCx+1.
ElbowPlot(spatial_6000, ndims = 40)
```

Based on this plot, we could roughly determine the majority of the variation by where the elbow occurs (touches the ground) to be between PC17-PC20. Determine percent of variation associated with each PC. While this gives us a good rough idea of the number of PCs needed to be included, a more quantitative approach may be a bit more reliable. We can calculate where the principal components start to elbow by taking the larger value of:

The point where the principal components only contribute 5% of standard deviation and the principal components cumulatively contribute 90% of the standard deviation.
The point where the percent change in variation between the consecutive PCs is less than 0.1%.
We will start by calculating the first metric:

```{r 6000firstmetric, echo=TRUE, warning=TRUE}
pct <- spatial_6000[["pca"]]@stdev / sum(spatial_6000[["pca"]]@stdev) * 100
# Calculate cumulative percents for each PC
cumu <- cumsum(pct)
# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
co1 <- which(cumu > 90 & pct < 5)[1]
co1
```
The first metric returns PC41 as the PC matching these requirements. Let’s check the second metric, which identifies the PC where the percent change in variation between consecutive PCs is less than 0.1%:

```{r 6000secondmetric, echo=TRUE, warning=TRUE}
# Determine the difference between variation of PC and subsequent PC
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1
# last point where change of % of variation is more than 0.1%.
co2
```
This second metric returns PC20. Usually, we would choose the minimum of these two metrics as the PCs covering the majority of the variation in the data.

```{r 6000minmetric, echo=TRUE, warning=TRUE}
# Minimum of the two calculation
pcs <- min(co1, co2)
pcs
```

```{r 6000clustering, echo=TRUE, warning=TRUE}
spatial_6000 <- FindNeighbors(spatial_6000, dims = 1:20)
spatial_6000 <- FindClusters(spatial_6000, verbose = FALSE)
spatial_6000 <- RunUMAP(spatial_6000, dims = 1:20)
```
```{r 6000multiresolution, echo=TRUE, warning=TRUE}
resolution_values <- c(0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3)
spatial_6000 <- FindClusters(spatial_6000, assay = "SCT", resolution = resolution_values)
plot <- clustree(spatial_6000, prefix = "SCT_snn_res.")
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/clustree.png", height = 3000, width = 3000, res = 300)
print(plot)
dev.off()
```

```{r 6000umapindividualsamplessave, echo=TRUE, warning=TRUE}
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = "SCT_snn_res.0.4", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/umappersample0.4.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = "SCT_snn_res.0.5", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/umappersample0.5.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = "SCT_snn_res.0.6", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/umappersample0.6.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = "SCT_snn_res.0.7", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3)
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/umappersample0.7.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = "SCT_snn_res.0.8", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/umappersample0.8.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = "SCT_snn_res.0.9", split.by = "orig.ident", label = TRUE, ncol = 3, label.size = 3) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/umappersample0.9.png", height = 3000, width = 2500, res = 300)
print(plot)
dev.off()
```
```{r 6000spatialumapindividualsamplessave, echo=TRUE, warning=TRUE}
plot <- SpatialDimPlot(spatial_6000, group.by = "SCT_snn_res.0.4", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/spatialumapres0.4.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_6000, group.by = "SCT_snn_res.0.5", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/spatialumapres0.5.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_6000, group.by = "SCT_snn_res.0.6", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/spatialumapres0.6.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_6000, group.by = "SCT_snn_res.0.7", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/spatialumapres0.7.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_6000, group.by = "SCT_snn_res.0.8", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/spatialumapres0.8.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
plot <- SpatialDimPlot(spatial_6000, group.by = "SCT_snn_res.0.9", label = TRUE, label.size = 3, ncol = 3)
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/spatialumapres0.9.png", height = 8000, width = 6000, res = 300)
print(plot)
dev.off()
```

```{r 6000pooledplot2, echo=TRUE, warning=TRUE}
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = c("SCT_snn_res.0.4", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/umap_res_0.4.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = c("SCT_snn_res.0.5", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/umap_res_0.5.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = c("SCT_snn_res.0.6", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/umap_res_0.6.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = c("SCT_snn_res.0.7", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/umap_res_0.7.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = c("SCT_snn_res.0.8", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/umap_res_0.8.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = c("SCT_snn_res.0.9", "orig.ident"), label = TRUE) 
plot
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/umap_res_0.9.png", height = 1500, width = 3500, res = 300)
print(plot)
dev.off()
```

```{r 6000pooledplot3, echo=TRUE, warning=TRUE}
plot <- DimPlot(spatial_6000, reduction = "umap", group.by = c("SCT_snn_res.0.4", "SCT_snn_res.0.5", "SCT_snn_res.0.6", "SCT_snn_res.0.7", "SCT_snn_res.0.8", "SCT_snn_res.0.9"), label = TRUE, label.size = 10, pt.size = 0.5, ncol = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/0.4to0.9_umap.png", height = 6000, width = 10000, res = 300)
print(plot)
dev.off()
plot <- DimPlot(spatial_6000, reduction = "pca", group.by = c("SCT_snn_res.0.4", "SCT_snn_res.0.5", "SCT_snn_res.0.6", "SCT_snn_res.0.7", "SCT_snn_res.0.8", "SCT_snn_res.0.9"), label = TRUE, label.size = 10, pt.size = 0.5, ncol = 3) 
png(filename = "../05.Analysis_in_R/3.Clustering/6000 genes 20 PCs used/0.4to0.9_pca.png", height = 6000, width = 10000, res = 300)
print(plot)
dev.off()
```



## Append Sample phenotype data to seurat object

```{r 3000_spatial_sampletypes, echo=TRUE, warning=TRUE}
# Assuming your Seurat object is named 'spatial_3000'
# Create a copy of the orig.ident column and rename it
spatial_3000$phenotype <- spatial_3000$orig.ident

Idents(spatial_3000) <- 'phenotype'
levels(spatial_3000)
spatial_3000 <- RenameIdents(spatial_3000,
"CR_1" = "CR",
"ACR_1" = "ACR",
"ACR_2" = "ACR",
"AKCR_1" = "early AKCR",
"AKCR_2" = "early AKCR",
"AKCR_3" = "early AKCR",
"AKCR_4" = "older AKCR",
"AKCR_5" = "older AKCR",
"AKCR_6" = "older AKCR",
"ACR_TU1" = "ACR tumor",
"ACR_TU2" = "ACR tumor",
"AKCR_TU1" = "AKCR tumor" )
levels(spatial_3000)
spatial_3000@meta.data$phenotype <- spatial_3000@active.ident
Idents(spatial_3000) <- 'SCT_snn_res.0.7'
head(spatial_3000)
# Remove the specified column (e.g., SCT_snn_res.0)
spatial_3000@meta.data$SCT_snn_res.0 <- NULL
spatial_3000@meta.data$SCT_snn_res.0.1 <- NULL
spatial_3000@meta.data$SCT_snn_res.0.2 <- NULL
spatial_3000@meta.data$SCT_snn_res.0.3 <- NULL
spatial_3000@meta.data$SCT_snn_res.0.4 <- NULL
spatial_3000@meta.data$SCT_snn_res.0.5 <- NULL
spatial_3000@meta.data$SCT_snn_res.0.6 <- NULL
spatial_3000@meta.data$SCT_snn_res.0.8 <- NULL
spatial_3000@meta.data$SCT_snn_res.0.9 <- NULL
spatial_3000@meta.data$SCT_snn_res.1 <- NULL
spatial_3000@meta.data$SCT_snn_res.1.1 <- NULL
spatial_3000@meta.data$SCT_snn_res.1.2 <- NULL
spatial_3000@meta.data$SCT_snn_res.1.3 <- NULL
spatial_3000@meta.data$seurat_clusters <- spatial_3000@meta.data$SCT_snn_res.0.7
head(spatial_3000)


# View the updated metadata
head(spatial_3000@meta.data)
```


